### PR TITLE
ContainerInterop compliance improvement

### PR DESCRIPTION
**Pull request by Sam-Burns**
Strict adherence to ContainerInterop standard w.r.t. exceptions

### CODE REVIEW

**Comment by Sam-Burns** [comment]
Referring to the [ContainerInterop specification on exceptions](https://github.com/container-interop/container-interop/blob/master/docs/ContainerInterface.md#12-exceptions)

File: Slim/Container.php [hunk 1]
```diff
@@ -263,7 +264,11 @@ public function get($id)
         if (!$this->offsetExists($id)) {
             throw new ContainerValueNotFoundException(sprintf('Identifier "%s" is not defined.', $id));
         }
-        return $this->offsetGet($id);
+        try {
```
Comment (geggleto) [inline Slim/Container.php:267]: The containers job is to retrieve the value and that's it. If the value throws an exception it's not the containers responsibility to deal with that.

**Comment by silentworks** [comment]
I think this part of the spec is a bit confusing, because the second sentence seems to call for the enforcement of the first sentence. 

Not sure why you would need both or why the container should be handling this part of exception handling rather than the end user doing it themselves.

//cc @mnapoli

**Comment by Sam-Burns** [comment]
The [docblocks](https://github.com/slimphp/Slim/blob/fb0e176a2ff6cea2fce0a946d3c60683e6d834c1/Slim/Container.php#L257) copied into Slim from the Interop `ContainerInterface` already suggest we are throwing an exception which complies with the standard, but none exists in the repo.

The intention behind having a consistent type of `ContainerException` is to allow the user clarity that it was specifically their DI config that caused the problem.

The PR goes further, adding the service ID to the message and making the previous exception accessible via `getPrevious()` - allowing easier debugging.

**Comment by mnapoli** [comment]
Hi!

Here is the spec:

> Exceptions directly thrown by the container MUST implement the Interop\Container\Exception\ContainerException.
> 
> A call to the get method with a non-existing id SHOULD throw a Interop\Container\Exception\NotFoundException.

Exceptions thrown by factories (i.e. closures) are not thrown by the container, so I don't see a reason to catch them and rethrow them as `ContainerException`.

> I think this part of the spec is a bit confusing, because the second sentence seems to call for the enforcement of the first sentence.

I'm not sure I understand what you mean. If you are saying that the second sentence conflicts with the first, I don't think it's the case because `NotFound` extends `ContainerException`.

**Comment by silentworks** [comment]
@mnapoli thanks for clearing that up. I didn't realise `NotFound` extended `ContainerException` hence why I found it a bit confusing.

**Comment by geggleto** [comment]
> Exceptions thrown by factories (i.e. closures) are not thrown by the container, so I don't see a reason to catch them and rethrow them as ContainerException.

@silentworks @mnapoli 

https://github.com/slimphp/Slim/pull/1663/files#diff-1fe12fa9ad45fa920bf98a8afcf0939eR270

This catches and rethrows the exceptions. I think we should remove this, thoughts?

**Comment by Sam-Burns** [comment]
I've rewritten it to take this into account.  Exceptions thrown explicitly by the container are now recast, but exceptions thrown in a factory closure are not.  This is captured by the unit tests added.

The only thing that can actually go wrong while retrieving a dependency (where `has()` would return `true`) is if there is a problem with the DI configuration.  At present, if a dependency factory closure tries to obtain another, non-existent, dependency from the container, it throws an `\InvalidArgumentException`.  The spec says that if the container explicitly throws an exception, it has to be an implementation of `ContainerException`, so something has to be done here.

File: Slim/Container.php [hunk 2]
```diff
@@ -8,12 +8,14 @@
  */
 namespace Slim;
 
+use Doctrine\Instantiator\Exception\InvalidArgumentException;
```
Comment (akrabat) [inline Slim/Container.php]: This is a dependency we would not want.

File: Slim/Container.php [hunk 3]
```diff
@@ -263,7 +265,30 @@ public function get($id)
         if (!$this->offsetExists($id)) {
             throw new ContainerValueNotFoundException(sprintf('Identifier "%s" is not defined.', $id));
         }
-        return $this->offsetGet($id);
+        try {
+            return $this->offsetGet($id);
+        } catch (\InvalidArgumentException $exception) {
+            if ($this->exceptionThrownByContainer($exception)) {
+                throw new SlimContainerException(
```
Comment (akrabat) [inline Slim/Container.php:271]: This is a BC break. Currently, calling code can expect to catch an `\InvalidArgumentException`, but this code changes that to a `\RuntimeException`.

Comment (Sam-Burns) [inline Slim/Container.php:271]: @akrabat The docblock says it is a `Interop\Container\Exception\ContainerException`.  What would your suggestion be?  Would you prefer to see `Slim\Exception\ContainerException` extend `\InvalidArgumentException` to preserve backwards compatibility?  I can change it if so.

Comment (mnapoli) [inline Slim/Container.php:271]: How about `Slim\Exception\ContainerEntryNotFoundException extends \InvalidArgumentException implements Interop\Container\NotFoundException` ? It's a "not found" kind of exception so it should extend `NotFoundException`.

Comment (silentworks) [inline Slim/Container.php:271]: But we already have a `ContainerValueNotFoundException` that extends `Interop\Container\NotFoundException` on line 266.

Comment (mnapoli) [inline Slim/Container.php:271]: Oh well then that's it!

Comment (akrabat) [inline Slim/Container.php:271]: > @akrabat The docblock says it is a Interop\Container\Exception\ContainerException. What would your suggestion be? Would you prefer to see Slim\Exception\ContainerException extend \InvalidArgumentException to preserve backwards compatibility? I can change it if so.

In this particular case, the only thing I especially care about is not changing anyone's code which calls get() within a try/catch block.

**Comment by Sam-Burns** [comment]
@akrabat I have changed the new exception class to preserve BC as discussed

**Comment by silentworks** [comment]
I am still not seeing the value in this, it feels like its been over-engineered for something that the developer should deal with themselves. Wouldn't this example below solve this issue?

``` php
$container = new Container;
$container['foo'] = function (ContainerInterface $container) {
    try {
        return $container['doesnt-exist'];
    } catch (\Exception $e) {
        // deal with it here
    }
};
$container->get('foo');
```

**Comment by Sam-Burns** [comment]
@silentworks The issue the class is trying to fix is that things like [this line](https://github.com/silexphp/Pimple/blob/v3.0.2/src/Pimple/Container.php#L96) in Pimple aren't compliant with the Container-Interop standard.  Wouldn't it be reasonable for a user to expect to be able to catch `Interop\Container\Exception\ContainerException` and handle or log that differently from other exceptions, if we're saying the container complies with the standard?

**Comment by silentworks** [comment]
Sorry for going back and forth on this one @Sam-Burns but we want to get to the best possible solution on this, I wonder if we should do as suggested on https://github.com/zendframework/zend-expressive/issues/129#issuecomment-157516173 until the actual Pimple team decides to adapt container interop.

**Comment by akrabat** [comment]
Where are we on this one?

**Comment by vlakoff** [comment]
One step forward from _Slim_ to [_Enterprise software_](https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition) (pun intended).

**Comment by vlakoff** [comment]
Seriously, IMHO, don't put band-aids like this just to forcibly adhere to some standard. It's a mess, just use Pimple as it is, it will adhere to the standard when time comes.

**Comment by mnapoli** [comment]
@vlakoff:
- Slim already supports container-interop, that's not the question here
- this issue is about minor improvements about how the standard is implemented
- if you want to use Pimple, you already can
- try to be more constructive and respectful of other's work

**Comment by vlakoff** [comment]
I just say catching exceptions and _trying_ to identify some of these to redefine them is not a good solution, IMHO.

I have nothing against this standard. It's just not implemented upstream (i.e. Pimple), so let's deal with it for now.

File: Slim/Container.php [hunk 4]
```diff
@@ -263,7 +264,30 @@ public function get($id)
         if (!$this->offsetExists($id)) {
             throw new ContainerValueNotFoundException(sprintf('Identifier "%s" is not defined.', $id));
         }
-        return $this->offsetGet($id);
+        try {
+            return $this->offsetGet($id);
+        } catch (\InvalidArgumentException $exception) {
+            if ($this->exceptionThrownByContainer($exception)) {
```
Comment (vlakoff) [inline Slim/Container.php:270]: If `(! $this->exceptionThrownByContainer($exception))` the exception is silently discarded, it should be re-thrown.

**Comment by akrabat** [comment]
Merged in 70c6d6d4c1b8c8d9beb788db438ee7442c2762cb.

Thanks for this work; I'm sorry it took so long to get in.

File: Slim/Container.php [hunk 5]
```diff
@@ -263,7 +264,30 @@ public function get($id)
...
+     * Tests whether an exception needs to be recast for compliance with Container-Interop.  This will be if the
+     * exception was thrown by Pimple.
+     *
+     * @param \InvalidArgumentException $exception
+     *
+     * @return bool
+     */
+    private function exceptionThrownByContainer(\InvalidArgumentException $exception)
+    {
+        return preg_match('/^Identifier ".*" is not defined.$/', $exception->getMessage());
```
Comment (vlakoff) [inline Slim/Container.php:290]: (nitpick) Final dot should be escaped `\.`

Comment (akrabat) [inline Slim/Container.php:290]: Good catch. Want to create a PR?

**Comment by vlakoff** [comment]
As I noted above, to not potentially lose the exception:

``` php
try {
    return $this->offsetGet($id);
} catch (\InvalidArgumentException $exception) {
    if ($this->exceptionThrownByContainer($exception)) {
        throw new SlimContainerException(...);
    }
    throw $exception; <-- this should be added
}
```

**Comment by akrabat** [comment]
Also good point. Not sure how I missed it. Please PR along with the escaped `.`

File: Slim/Container.php (hunk 5 above)
Comment (vlakoff) [inline Slim/Container.php]: I think that would be overkill.

**Comment by vlakoff** [comment]
Maybe the trace could be used to make the detection robuster?

As a quick example:

```
$trace = $exception->getTrace();
$trace[0]['class'] === 'Pimple\Container';
$trace[0]['function'] === 'offsetGet';
```

**Pull request closed without merging**